File: LUT_COS/lut_cos.py
import math
import numpy as np
import matplotlib.pyplot as plt

LUT=[
     1.0000, 0.9848,  0.9396, 0.8660,
     0.7660, 0.6427,  0.5000, 0.3420,
     0.1736, 0.0000, -0.1736,-0.3420,
    -0.5000,-0.6427, -0.7660,-0.8660,
    -0.9396,-0.9848, -1.0000,-0.9848,
    -0.9396,-0.8660, -0.7660,-0.6427,
    -0.5000,-0.3420, -0.1736,-0.0000,
     0.1736, 0.3420,  0.5000, 0.6427,
     0.7660, 0.8660,  0.9396, 0.9848, 1.0]

def lut_cos_deg(x):
    x%=360
    u=x//10
    v=x%10
    a,b=LUT[u],LUT[u+1]
    return a+(b-a)*float(v)/10.0

x =np.array([np.cos(math.radians(d)) for d in range(360)])
x0=np.array([lut_cos_deg(d)          for d in range(360)])

plt.clf()
plt.plot(x,'--k')
plt.plot(x0,'k')
plt.grid(True)
plt.show()
    
plt.plot(np.array(x)-np.array(x0),'k')
plt.grid(True)
plt.show()

File: LUT_COS/lut_cos.c
#include <stdio.h>

#define _USE_MATH_DEFINES
#include <math.h>

const float LUT[]=
{
     1.0000f, 0.9848f,  0.9396f, 0.8660f,
     0.7660f, 0.6427f,  0.5000f, 0.3420f,
     0.1736f, 0.0000f, -0.1736f,-0.3420f,
    -0.5000f,-0.6427f, -0.7660f,-0.8660f,
    -0.9396f,-0.9848f, -1.0000f,-0.9848f,
    -0.9396f,-0.8660f, -0.7660f,-0.6427f,
    -0.5000f,-0.3420f, -0.1736f,-0.0000f,
     0.1736f, 0.3420f,  0.5000f, 0.6427f,
     0.7660f, 0.8660f,  0.9396f, 0.9848f, 1.0f
};

float lut_cos_deg(unsigned x)
{
    x%=360;
    unsigned u=x/10,v=x%10;
    float a=LUT[u], b=LUT[u+1];
    return (float)(a+(b-a)*((float)v)/10.0f);
}

int main()
{
    float err=0, y0, y1;
    for (unsigned x=0; x<360; x++)
    {
        y0=(float)cos((float)x*M_PI/180.0f);
        y1=lut_cos_deg(x);
        float e=fabsf(y0-y1);
        if (e>err)
            err=e;
    }
    printf("[INF] error: %e\n",err);
}
